lib/OrdinaryDiffEqCore/src/integrators/controllers.jl

```diff
@@ -333,10 +511,139 @@ function step_accept_controller!(integrator, controller::PIDController, alg, dt_
+    qmin::T
+    qmax::T
+    gamma::T
+    qsteady_min::T
+    qsteady_max::T
+    qoldinit::T
+end
+
+function PIDController(alg; kwargs...)
+    PIDController(Float64, alg; kwargs...)
+end
+
+function PIDController(QT, alg; beta1 = nothing, beta2 = nothing, beta3 = nothing, accept_safety = 0.81, limiter = default_dt_factor_limiter, qmin = nothing, qmax = nothing, gamma = nothing, qsteady_min = nothing, qsteady_max = nothing, qoldinit = nothing)
+    beta2 = beta2 === nothing ? beta2_default(alg) : beta2
+    beta1 = beta1 === nothing ? beta1_default(alg, beta2) : beta1
+    beta3 = beta3 === nothing ? zero(beta1) : beta3
+    beta = MVector(map(float, promote(beta1, beta2, beta3))...)
+    PIDController{QT}(
+        beta,
+        err,
+        accept_safety,
+        limiter,
+        qmin === nothing ? qmin_default(alg) : qmin,
+        qmax === nothing ? qmax_default(alg) : qmax,
+        gamma === nothing ? gamma_default(alg) : gamma,
+        qsteady_min === nothing ? qsteady_min_default(alg) : qsteady_min,
+        qsteady_max === nothing ? qsteady_max_default(alg) : qsteady_max,
+        qoldinit === nothing ? 1 // 10^4 : qoldinit,
+    )
+end
+
+function Base.show(io::IO, controller::PIDController)
+    print(io, "PIDController(beta=", controller.beta,
+        ", accept_safety=", controller.accept_safety,
+        ", limiter=", controller.limiter,
+        ")")
+end
+
+struct PIDControllerCache{T, Limiter} <: AbstractControllerCache
+    controller::PIDController{T, Limiter}
+    err::MVector{3, T} # history of the error estimates
+    dt_factor::T
+end
+
+function setup_controller_cache(alg_cache, controller::PIDController{QT}) where QT
+    err = MVector{3, QT}(true, true, true)
+    PIControllerCache(
+        controller,
+        err,
+        QT(1 // 10^4),
+    )
+end
+
+@inline function stepsize_controller!(integrator, cache::PIDControllerCache, alg)
+    @unpack controller = cache
+    beta1, beta2, beta3 = controller.beta
+    @unpack qmax = controller
+
+    EEst = DiffEqBase.value(integrator.EEst)
+
+    # If the error estimate is zero, we can increase the step size as much as
+    # desired. This additional check fixes problems of the code below when the
+    # error estimates become zero
+    # -> err1, err2, err3 become Inf
+    # -> err1^positive_number * err2^negative_number becomes NaN
+    # -> dt becomes NaN
+    #
+    # `EEst_min` is smaller than PETSC_SMALL used in the equivalent logic in PETSc.
+    # For example, `eps(Float64) ≈ 2.2e-16` but `PETSC_SMALL ≈ 1.0e-10` for `double`.
+    EEst_min = eps(typeof(EEst))
+    # The code below is a bit more robust than
+    # ```
+    # if iszero(EEst)
+    #   EEst = eps(typeof(EEst))
+    # end
+    # ```
+    EEst = ifelse(EEst > EEst_min, EEst, EEst_min)
+
+    cache.err[1] = inv(EEst)
+    err1, err2, err3 = cache.err
+
+    k = min(alg_order(alg), alg_adaptive_order(alg)) + 1
+    dt_factor = err1^(beta1 / k) * err2^(beta2 / k) * err3^(beta3 / k)
```

Comment:
use `fastpow` and `inv(k)` to save computation

Reply:
it looks like the differences between FastPow and ^ are basically noise (since timestepping a chaotic system will take a pseudorandom number of steps depending on epsilon level differences). This PR doesn't change the code. Dennis and I have discussed a followup PR with some further cleanup, but this is pure refactoring.